anvil/models.py

```diff
@@ -240,7 +234,17 @@ def log_det_jacobian(self, phi_input):
         """
         a_for_net = phi_input[:, self._a_ind]  # select phi_a
         s_out = self._s_forward(a_for_net)
-        return s_out.sum(dim=-1)
+        return s_out.sum(dim=1, keepdim=True)
+
+    def forward(self, z_input: torch.Tensor) -> tuple:
+        """Given an stack of states z_input, returns a tuple 
+        (phi_out, log_det_jacob) containing the (inversely) transformed
+        states and the log of the jacobian determinant of the forward
+        transformation.
+        """
+        phi_out = self.inverse_coupling_layer(z_input)
+        log_det_jacob = self.log_det_jacobian(phi_out)
+        return phi_out, log_det_jacob
```

Comment:
@wilsonmr is this roughly what you had in mind? (but with better docstring)

Reply:
the output is good yeah. Now please check this because if we get this wrong we break the whole thing lol

since phi = [phi_a, phi_b] -> inverse coupling -> [phi_a, z_b] (in other words phi_a is always left untouched by transformation)

then I think that if we had

```python
    def forward(self, z_input):
        z_a = z_input[:, self._a_ind]
        z_b = z_input[:, self._b_ind]
        s_out = self._s_forward(z_a)
        t_out = self._t_forward(z_a)
        phi_b = (z_b - t_out) * torch.exp(-s_out)
        return self.join_func([z_a, phi_b], dim=1), s_out.sum(dim=1, keepdims=True)
```

then we would get the same thing, with one less pass through the net.

I made a dummy_forward which was a copy paste of the above and it seems fine to me:

```python
>>> l = AffineLayer(10, (10, 10), (10, 10), 0)
>>> 
>>> import torch
>>> test_tensor = torch.rand(1, 10)
>>> 
>>> l(test_tensor)
(tensor([[0.9308, 0.7645, 0.3430, 0.7738, 0.1795, 1.0089, 0.7677, 0.5743, 0.4620,
         0.2349]], grad_fn=<CatBackward>), tensor([[0.1624]], grad_fn=<SumBackward1>))
>>> l.dummy_forward(test_tensor)
(tensor([[0.9308, 0.7645, 0.3430, 0.7738, 0.1795, 1.0089, 0.7677, 0.5743, 0.4620,
         0.2349]], grad_fn=<CatBackward>), tensor([[0.1624]], grad_fn=<SumBackward1>))
```

and we see it's faster

```python
In [1]: from anvil.models import AffineLayer                                                                                                                                                                

In [2]: import torch                                                                                                                                                                                        

In [3]: tens = torch.rand(1, 10)                                                                                                                                                                            

In [4]: l = AffineLayer(10, (10, 10), (10, 10), 0)                                                                                                                                                          

In [5]: def test_dummy(s, lay): 
   ...:     with torch.no_grad(): 
   ...:         return lay.dummy_forward(s) 
   ...:                                                                                                                                                                                                     

In [6]: def test_forward(s, lay): 
   ...:     with torch.no_grad(): 
   ...:         return lay(s) 
   ...:                                                                                                                                                                                                     

In [7]: %timeit test_dummy(tens, l)                                                                                                                                                                         
135 µs ± 359 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [8]: %timeit test_forward(tens, l)                                                                                                                                                                       
192 µs ± 848 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

and remember that speed up scales with size of the network!

Reply:
ah great - more gains. I guess we're rediscovering the USP of RealNVP which is that it's dirt cheap.

I will update and run some checks.

Reply:
oops @ keepdims

Reply:
they purposely made tensors behave like numpy arrays, not sure why they changed all of the nomenclature!

Reply:
@marshrossney did you change anything here? I don't see it